src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentsService.java

```diff
@@ -79,28 +89,42 @@ public Map<Locale, Map<String, Map<String, String>>> getTranslation(final Set<St
-		final JSONArray bundles = (JSONArray) opt.getDataPart(response).get(ConstantsKeys.BUNDLES);
-        final JSONArray localesFromServer = (JSONArray) opt.getDataPart(response).get(ConstantsKeys.LOCALES);
-        final Map<String, String> localeMap = this.makeLocaleMap(localesToQuery, localesFromServer);
-
-        // combine data from server into the map to return.
-        final Iterator<?> iter = bundles.iterator();
-        while (iter.hasNext()) {
-            final JSONObject bundle = (JSONObject) iter.next();
-            final String locale = localeMap.get(bundle.get(ConstantsKeys.LOCALE));
-            final String comp = (String) bundle.get(ConstantsKeys.COMPONENT);
-            final JSONObject messages = (JSONObject) bundle.get(ConstantsKeys.MESSAGES);
-
-            // update cache.
-            final MessagesDTO dto = new MessagesDTO();
-            dto.setComponent(comp);
-            dto.setLocale(locale);
-            new CacheService(dto).addCacheOfComponent(new MessageCacheItem(messages, null,
-            		cacheItem.getTimestamp(), cacheItem.getMaxAgeMillis()));
-
-            // update map to return.
-            dataMap.get(locale).put(comp, messages);
+        JSONObject data = opt.queryFromServer(componentsToQuery, localesToQuery, cacheItem);
+        if(data != null) {
+            final JSONArray bundles = (JSONArray) data.get(ConstantsKeys.BUNDLES);
+            final JSONArray localesFromServer = (JSONArray) data.get(ConstantsKeys.LOCALES);
+            final Map<String, String> localeMap = this.makeLocaleMap(localesToQuery, localesFromServer);
+
+            final SortedSet<String> localesToFallback = new TreeSet<>();
+            final SortedSet<String> componentsToFallback = new TreeSet<>();
+            // combine data from server into the map to return.
+            final Iterator<?> iter = bundles.iterator();
+            while (iter.hasNext()) {
+                final JSONObject bundle = (JSONObject) iter.next();
+                final String locale = localeMap.get(bundle.get(ConstantsKeys.LOCALE));
+                final String comp = (String) bundle.get(ConstantsKeys.COMPONENT);
+                final JSONObject messages = (JSONObject) bundle.get(ConstantsKeys.MESSAGES);
+
+                if (messages != null && !"".equals(messages)) {
+                    // update cache.
+                    final MessagesDTO dto = new MessagesDTO();
+                    dto.setComponent(comp);
+                    dto.setLocale(locale);
+                    new CacheService(dto).addCacheOfComponent(new MessageCacheItem(messages, null,
+                            cacheItem.getTimestamp(), cacheItem.getMaxAgeMillis()));
+
+                    // update map to return.
+                    dataMap.get(locale).put(comp, messages);
+                } else {
+                    localesToFallback.add(locale);
+                    componentsToFallback.add(comp);
```

Comment:
why not get the messages of the locale/component here and update the cache?

Reply:
I meant put the codes of function getFallbackLocaleMessages at here, because it makes duplicated to loop the locales and components again by localesToFallback and componentsToFallback.

Reply:
The codes of function getFallbackLocaleMessages can't be put here.
One reason is here only can do one locale-component's fallback since it's in the loop, but getFallbackLocaleMessages do multiple locale components fallback; another reason is this method is already about 100 lines, it put the codes here, this method will be harder to read and more complex.

Reply:
the point is the loop is duplicated, also more extra codelines are introduced becasue of 'localesToFallback' and 'componentsToFallback'.

Reply:
I got your point. But I think in most cases the size of 'localesToFallback' and 'componentsToFallback' are not big, the loop of them won't take much time and impact performance.